chalk-solve/src/infer/unify.rs

```diff
@@ -274,38 +383,356 @@ impl<'t, I: Interner> Unifier<'t, I> {
         // for<'a...> exists<'b...> T == U &&
         // for<'b...> exists<'a...> T == U
 
+        // for<'a...> T <: for<'b...> U
+        //
+        // if
+        //
+        // for<'b...> exists<'a...> T <: U
+
         let interner = self.interner;
 
-        {
+        if let Variance::Invariant | Variance::Contravariant = variance {
             let a_universal = self.table.instantiate_binders_universally(interner, a);
             let b_existential = self.table.instantiate_binders_existentially(interner, b);
-            Zip::zip_with(self, &a_universal, &b_existential)?;
+            Zip::zip_with(self, variance, &a_universal, &b_existential)?;
```

Comment:
```suggestion
            Zip::zip_with(self, Variance::Contravariant, &a_universal, &b_existential)?;
```
(And `Covariant` below). This is required for the tests I've commented on to give the correct results.

Reply:
*Is* `Contravariant` here right? I don't think so, since this also covers the `Invariant` case.

Reply:
In the invariant case this `if` uses `Contravariant` and the other one uses `Covariant`. This is required so that `T == U` is the same as `T <: U` and `U <: T`.